### DISCUSSION TITLE
craft\debug\RequestPanel::save modify `actionParams` key to prevent `yii\debug\FlattenException` when `Yii::$app->requestedParams` contains Closures

### DISCUSSION THREAD

**Discussion by Anubarak** [craftcms/cms#14982] [Ideas] [answered]
Currently Yii2 Debug Toolbar logs the entire `Yii::$app->requestedParams` in their Request Panel.

This is usually quite good, but sometimes we use `Behaviors` or `Events` as Closures for the current `Yii::$app->requestedParams`.

Let's say we have a site request for an entry. In that case
`Yii::$app->requestedParams['variables']['entry']` is a very huge array with all behaviours and events. An Example can be found in this dump
[actionParams.txt](https://github.com/craftcms/cms/files/15291679/actionParams.txt)

This Entry has not yet loaded relation fields:
```
["contentMatrix"]=>
object(craft\elements\db\EntryQuery)#765 (100) {
["_events":"yii\base\Component":private]=>
array(1) {
  ["beforePrepare"]=>
  array(1) {
	[0]=>
	array(2) {
	  [0]=>
	  object(Closure)#772 (3) {
		["static"]=>
		array(1) {
		  ["handler"]=>
		  object(Closure)#770 (3) {
			["static"]=>
			array(1) {
			  ["owner"]=>
			  *RECURSION*
			}
			["this"]=>
// snip
```
Unfortunately Closures cannot be serialized thus the data cannot be rendered and a FlattenArrayException is thrown
https://github.com/yiisoft/yii2-debug/blob/2.1.1/src/LogTarget.php#L60-L62

What I usually do is to overwrite your `craft\debug\RequestPanel` to unset `actionParams` since I usually do not need them, but maybe you could include a function that filters out at least ElementInterfaces or at least Behaviours or Events of those. As of right now most normal template requests  with additional events attached to Entries are unable to render.

```PHP
class RequestPanel extends \craft\debug\RequestPanel
{
    public function save(): array
    {
        $save = parent::save();

        if(isset($save['actionParams']) && \is_array($save['actionParams'])){
            $save['actionParams'] = $this->filterParams($save['actionParams']);
        }
        return $save;
    }

    protected function filterParams(array $params): array
    {
        foreach ($params as $key => $value){
            if($value instanceof ElementInterface){
                $params[$key] = $value::class . '::' . $value->getRef();
            } else if(is_array($value)){
                $params[$key] = $this->filterParams($value);
            } else if(\is_object($value)){
                $params[$key] = $value::class;
            } else {
                $params[$key] = $value;
            }
        }

        return $params;
    }
}
```

**Comment by brandonkelly** [comment, marked as answer]
Thanks for pointing that out!

I’ve updated `RequestPanel` to start pre-serializing any objects within `actionParams` via `VarDumper::dump()`, so `serialize()` doesn’t potentially throw an error due to closures. (44842162f5655628ead94d9c052f7c2ecef64c23)

The change was a little involved, so I made it for 4.10 and 5.2.

![The “Parameters” output in the Debug Toolbar’s Request panel, showing that closures have been serialized into “Closure” output.](https://github.com/craftcms/cms/assets/47792/7e1b711a-cfe8-4bcd-abf7-3055b4cadfdc)

**Reply by brandonkelly** [reply]
Craft 4.10.0-beta.1 and 5.2.0-beta.1 are out with that change.
